node/cmd/ccq/http.go

```diff
@@ -0,0 +1,143 @@
+package ccq
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sort"
+	"time"
+
+	gossipv1 "github.com/certusone/wormhole/node/pkg/proto/gossip/v1"
+	"github.com/gorilla/mux"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"google.golang.org/protobuf/proto"
+)
+
+type queryRequest struct {
+	Bytes     string `json:"bytes"`
+	Signature string `json:"signature"`
+}
+
+type queryResponse struct {
+	Bytes      string   `json:"bytes"`
+	Signatures []string `json:"signatures"`
+}
+type errorResponse struct {
+	Error string `json:"error"`
+}
+
+type httpServer struct {
+	topic            *pubsub.Topic
+	pendingResponses *PendingResponses
+}
+
+func (s *httpServer) handleQuery(w http.ResponseWriter, r *http.Request) {
+	enc := json.NewEncoder(w)
+
+	var q queryRequest
+	err := json.NewDecoder(r.Body).Decode(&q)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		enc.Encode(errorResponse{Error: err.Error()})
+		return
+	}
+
+	queryRequestBytes, err := hex.DecodeString(q.Bytes)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		enc.Encode(errorResponse{Error: err.Error()})
+		return
+	}
+
+	// TODO: check if request signer is authorized on Wormchain
+
+	signature, err := hex.DecodeString(q.Signature)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		enc.Encode(errorResponse{Error: err.Error()})
+		return
+	}
+
+	signedQueryRequest := &gossipv1.SignedQueryRequest{
+		QueryRequest: queryRequestBytes,
+		Signature:    signature,
+	}
+
+	// TODO: validate request before publishing
+
+	m := gossipv1.GossipMessage{
+		Message: &gossipv1.GossipMessage_SignedQueryRequest{
+			SignedQueryRequest: signedQueryRequest,
+		},
+	}
+
+	b, err := proto.Marshal(&m)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		enc.Encode(errorResponse{Error: err.Error()})
+		return
+	}
+
+	pendingResponse := NewPendingResponse(signedQueryRequest)
+	added := s.pendingResponses.Add(pendingResponse)
+	if !added {
+		w.WriteHeader(http.StatusInternalServerError)
+		enc.Encode(errorResponse{Error: "Duplicate request"})
+		return
+	}
+
+	err = s.topic.Publish(r.Context(), b)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		s.pendingResponses.Remove(pendingResponse)
+		return
+	}
+
+	// Wait for the response or timeout
+	select {
+	case <-time.After(1 * time.Minute):
```

Comment:
The query handler waits a minute. Maybe this should wait a bit longer to allow for roundtrip over gossip?

PR #3112 makes `query.RequestTime` available. Maybe you could use that plus some number of seconds?

Reply:
I think for now we could simply wait an additional 5 seconds for the message to propagate